File: src/main/java/aoc2019/Day19.java
package aoc2019;

import java.util.Arrays;
import java.util.List;

public class Day19 {

    public static void main(String[] args) {

        long[] array = Arrays.stream(
                "109,424,203,1,21102,11,1,0,1105,1,282,21102,1,18,0,1106,0,259,2101,0,1,221,203,1,21102,1,31,0,1106,0,282,21102,38,1,0,1105,1,259,20101,0,23,2,22101,0,1,3,21101,1,0,1,21101,57,0,0,1105,1,303,2101,0,1,222,21001,221,0,3,21002,221,1,2,21101,0,259,1,21102,80,1,0,1106,0,225,21102,89,1,2,21102,91,1,0,1105,1,303,2101,0,1,223,20101,0,222,4,21101,0,259,3,21102,1,225,2,21102,225,1,1,21102,118,1,0,1106,0,225,20101,0,222,3,21101,136,0,2,21101,133,0,0,1106,0,303,21202,1,-1,1,22001,223,1,1,21101,148,0,0,1105,1,259,1202,1,1,223,20102,1,221,4,21001,222,0,3,21102,18,1,2,1001,132,-2,224,1002,224,2,224,1001,224,3,224,1002,132,-1,132,1,224,132,224,21001,224,1,1,21102,195,1,0,106,0,108,20207,1,223,2,20102,1,23,1,21101,-1,0,3,21101,214,0,0,1105,1,303,22101,1,1,1,204,1,99,0,0,0,0,109,5,1202,-4,1,249,21201,-3,0,1,22102,1,-2,2,21202,-1,1,3,21102,1,250,0,1105,1,225,21201,1,0,-4,109,-5,2105,1,0,109,3,22107,0,-2,-1,21202,-1,2,-1,21201,-1,-1,-1,22202,-1,-2,-2,109,-3,2105,1,0,109,3,21207,-2,0,-1,1206,-1,294,104,0,99,22102,1,-2,-2,109,-3,2105,1,0,109,5,22207,-3,-4,-1,1206,-1,346,22201,-4,-3,-4,21202,-3,-1,-1,22201,-4,-1,2,21202,2,-1,-1,22201,-4,-1,1,21201,-2,0,3,21102,343,1,0,1106,0,303,1105,1,415,22207,-2,-3,-1,1206,-1,387,22201,-3,-2,-3,21202,-2,-1,-1,22201,-3,-1,3,21202,3,-1,-1,22201,-3,-1,2,21202,-4,1,1,21102,384,1,0,1105,1,303,1106,0,415,21202,-4,-1,-4,22201,-4,-3,-4,22202,-3,-2,-2,22202,-2,-4,-4,22202,-3,-2,-3,21202,-4,-1,-2,22201,-3,-2,1,21202,1,1,-4,109,-5,2106,0,0".split(
                        ",")).mapToLong(Long::parseLong).toArray();

        int part1 = 0;
        for (int x = 0; x < 50; x++) {
            for (int y = 0; y < 50; y++) {
                if (inRange(x, y, array)) {
                    part1++;
                }
            }
        }
        System.out.println("Part 1: " + part1);


        int x = 0;
        int y = 10;

        while (true) {
            if (inRange(x, y, array)) {
                if (inRange(x, y, array)) {
                    break;
                }
                y++;
            } else {
                x++;
            }
        }
        int part2 = 10_000 * x + y - 99;
        System.out.println("Part 2: " + part2);
    }

    private static boolean inRange(int x, int y, long[] array) {
        return IntCode.run(array, List.of(x, y).iterator()) == 1;
    }
}

File: src/main/java/aoc2021/Day19.kt
package aoc2021

import java.util.*
import kotlin.math.abs

fun main() {
    Day19().run()
}

class Day19 : Day(19) {
    override fun part1(input: List<String>): Any {
        val scanners =
            ArrayList(input.joinToString("\n").split(Regex("-{3} scanner \\d+ -{3}"))
                .filter(String::isNotEmpty).map(::Scanner))
        val allBeacons = HashSet(scanners.removeFirst().relativeBeacons)
        while (scanners.isNotEmpty()) {
            scanners.removeIf { attemptAlign(allBeacons, it) != null }
        }
        return allBeacons.size
    }

    override fun part2(input: List<String>): Any {
        val scanners =
            ArrayList(input.joinToString("\n").split(Regex("-{3} scanner \\d+ -{3}"))
                .filter(String::isNotEmpty).map(::Scanner))
        val allBeacons = HashSet(scanners.removeFirst().relativeBeacons)
        val points = HashSet<Point>()
        points.add(Point(0, 0, 0))
        while (scanners.isNotEmpty()) {
            scanners.removeIf {
                val point = attemptAlign(allBeacons, it)
                if (point != null)
                    points.add(point)
                point != null
            }
        }
        return points.flatMap { points.map(it::manhattan) }.maxOrNull()!!
    }

    private fun attemptAlign(allBeacons: MutableCollection<Point>, scanner: Scanner): Point? {
        val commonDist =
            allBeacons.pairDist().filter { (_, v) -> scanner.pairDist.containsValue(v) }
                .mapValues { (_, v) ->
                    scanner.pairDist.filter { it.value == v }.map(
                        Map.Entry<Pair<Point, Point>, Point>::key)[0]

                }
        if (commonDist.isEmpty())
            return null
        val (pair1, pair2) = commonDist.entries.random()
        try {
            val absolute = pair1.first
            val relative = pair2.first
            val (mapping, position) =
                rotation.mapValues { absolute - it.value(relative) }.entries
                    .filter { (mapping, position) ->
                        commonDist.all { (k, v) ->
                            val mapped = mapping(k.first - position)
                            mapped == v.first || mapped == v.second
                        }
                    }[0]
            allBeacons.addAll(scanner.relativeBeacons.map { rotation[mapping]!!(it) + position })
            return position

        } catch (e: IndexOutOfBoundsException) {
            val absolute = pair1.first
            val relative = pair2.second
            val (mapping, position) =
                rotation.mapValues { absolute - it.value(relative) }.entries
                    .filter { (mapping, position) ->
                        commonDist.all { (k, v) ->
                            val mapped = mapping(k.first - position)
                            mapped == v.first || mapped == v.second
                        }
                    }[0]
            allBeacons.addAll(scanner.relativeBeacons.map { rotation[mapping]!!(it) + position })
            return position
        }
    }

    private fun Iterable<Point>.pairDist(): Map<Pair<Point, Point>, Point> {
        val pairDist: MutableMap<Pair<Point, Point>, Point> = HashMap()
        for (b1 in this) {
            for (b2 in this) {
                if (!pairDist.contains(Pair(b2, b1)) && b1 != b2)
                    pairDist[Pair(b1, b2)] = (b1 - b2).hash()
            }
        }
        return pairDist
    }


   private class Scanner(val relativeBeacons: List<Point>) {

        val pairDist: MutableMap<Pair<Point, Point>, Point> = HashMap()

        init {
            for (b1 in relativeBeacons) {
                for (b2 in relativeBeacons) {
                    if (!pairDist.contains(Pair(b2, b1)) && b1 != b2)
                        pairDist[Pair(b1, b2)] = (b1 - b2).hash()
                }
            }
        }

        constructor(input: String) : this(input.split("\n").filter(String::isNotEmpty)
            .map(::Point))
    }

    @SuppressWarnings("unused")
    private data class Point(val x: Int, val y: Int, val z: Int) {

        constructor(list: List<Int>) : this(list[0], list[1], list[2])

        constructor(line: String) : this(line.split(',').filter(String::isNotEmpty)
            .map(String::toInt))


        operator fun plus(p: Point) = Point(x + p.x, y + p.y, z + p.z)

        operator fun minus(p: Point) = Point(x - p.x, y - p.y, z - p.z)

        infix fun manhattan(p: Point) = abs(x - p.x) + abs(y - p.y) + abs(z - p.z)

        fun hash() = Point(listOf(x, y, z).map(::abs).sorted())

    }

    private val rotation: Map<(Point) -> Point, (Point) -> Point> = mapOf(
        Pair({ (x, y, z) -> Point(x, y, z) }) { (x, y, z) -> Point(x, y, z) },
        Pair({ (x, y, z) -> Point(y, z, x) }) { (x, y, z) -> Point(z, x, y) },
        Pair({ (x, y, z) -> Point(z, x, y) }) { (x, y, z) -> Point(y, z, x) },
        Pair({ (x, y, z) -> Point(-x, z, y) }) { (x, y, z) -> Point(-x, z, y) },
        Pair({ (x, y, z) -> Point(z, y, -x) }) { (x, y, z) -> Point(-z, y, x) },
        Pair({ (x, y, z) -> Point(y, -x, z) }) { (x, y, z) -> Point(-y, x, z) },
        Pair({ (x, y, z) -> Point(x, z, -y) }) { (x, y, z) -> Point(x, -z, y) },
        Pair({ (x, y, z) -> Point(z, -y, x) }) { (x, y, z) -> Point(z, -y, x) },
        Pair({ (x, y, z) -> Point(-y, x, z) }) { (x, y, z) -> Point(y, -x, z) },
        Pair({ (x, y, z) -> Point(x, -z, y) }) { (x, y, z) -> Point(x, z, -y) },
        Pair({ (x, y, z) -> Point(-z, y, x) }) { (x, y, z) -> Point(z, y, -x) },
        Pair({ (x, y, z) -> Point(y, x, -z) }) { (x, y, z) -> Point(y, x, -z) },
        Pair({ (x, y, z) -> Point(-x, -y, z) }) { (x, y, z) -> Point(-x, -y, z) },
        Pair({ (x, y, z) -> Point(-y, z, -x) }) { (x, y, z) -> Point(-z, -x, y) },
        Pair({ (x, y, z) -> Point(z, -x, -y) }) { (x, y, z) -> Point(-y, -z, x) },
        Pair({ (x, y, z) -> Point(-x, y, -z) }) { (x, y, z) -> Point(-x, y, -z) },
        Pair({ (x, y, z) -> Point(y, -z, -x) }) { (x, y, z) -> Point(-z, x, -y) },
        Pair({ (x, y, z) -> Point(-z, -x, y) }) { (x, y, z) -> Point(-y, z, -x) },
        Pair({ (x, y, z) -> Point(x, -y, -z) }) { (x, y, z) -> Point(x, -y, -z) },
        Pair({ (x, y, z) -> Point(-y, -z, x) }) { (x, y, z) -> Point(z, -x, -y) },
        Pair({ (x, y, z) -> Point(-z, x, -y) }) { (x, y, z) -> Point(y, -z, -x) },
        Pair({ (x, y, z) -> Point(-x, -z, -y) }) { (x, y, z) -> Point(-x, -z, -y) },
        Pair({ (x, y, z) -> Point(-z, -y, -x) }) { (x, y, z) -> Point(-z, -y, -x) },
        Pair({ (x, y, z) -> Point(-y, -x, -z) }) { (x, y, z) -> Point(-y, -x, -z) }
    )


}




